File: 4/android/temu16/fbaseVideo/app/src/main/java/com/khenjy/fbasevideo/teachers/MainActivity.java
package com.khenjy.fbasevideo.teachers;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.SearchView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.firebase.ui.database.FirebaseRecyclerOptions;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.khenjy.fbasevideo.R;
import com.orhanobut.dialogplus.DialogPlus;
import com.orhanobut.dialogplus.ViewHolder;

public class MainActivity extends AppCompatActivity {

    RecyclerView recyclerView;
    MainAdapter mainAdapter;
    FloatingActionButton floatingActionButton;
    private static final int BACK_PRESS_INTERVAL = 2000; // 2 seconds
    private long backPressTime;


    FirebaseDatabase database = FirebaseDatabase.getInstance();
    DatabaseReference myRef = database.getReference("message");

    @Override
    public void onBackPressed() {
        if (backPressTime + BACK_PRESS_INTERVAL > System.currentTimeMillis()) {
            super.onBackPressed();
            finishAffinity(); // Closes the app completely
        } else {
            Toast.makeText(this, "Press back again to exit.", Toast.LENGTH_SHORT).show();
        }
        backPressTime = System.currentTimeMillis();
    }

    Button btnAbout; // Initialization moved here

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.teachers_activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        // Initialization moved after setContentView
        btnAbout = findViewById(R.id.btnAbout);

        myRef.child("teachers").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull Task<DataSnapshot> task) {
                if (task.isSuccessful()) {
                    Log.e("firebase", "Error Getting Data", task.getException());
                } else {
                    Log.d("firebase", "Data: " + task.getResult().getValue());
                }
            }
        });

        recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        FirebaseRecyclerOptions<MainModel> options =
                new FirebaseRecyclerOptions.Builder<MainModel>()
                        .setQuery(FirebaseDatabase.getInstance().getReference().child("teachers"), MainModel.class)
                        .build();

        mainAdapter = new MainAdapter(options);
        recyclerView.setAdapter(mainAdapter);

        floatingActionButton = findViewById(R.id.floatingActionButton);
        floatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent move = new Intent(MainActivity.this, AddActivity.class);
                startActivity(move);
            }
        });

        btnAbout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final DialogPlus dialogPlus = DialogPlus.newDialog(MainActivity.this)
                        .setContentHolder(new ViewHolder(R.layout.teachers_about_popup))
                        .setExpanded(true, 2000)
                        .create();

                //dialogPlus.show();

                View view = dialogPlus.getHolderView();

                dialogPlus.show();
            }
        });

        // Register the view (e.g., a TextView) for which you want to show the context menu
    }

    @Override
    protected void onStart() {
        super.onStart();
        mainAdapter.startListening();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mainAdapter.stopListening();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.search, menu);

        // Associate searchable configuration with the SearchView
        MenuItem item = menu.findItem(R.id.search);
        SearchView searchView = (SearchView) item.getActionView();

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                txtSearch(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String query) {
                txtSearch(query);
                return false;
            }
        });
        return super.onCreateOptionsMenu(menu);

    }

    private void txtSearch(String str) {
        FirebaseRecyclerOptions<MainModel> options =
                new FirebaseRecyclerOptions.Builder<MainModel>()
                        .setQuery(FirebaseDatabase.getInstance().getReference().child("teachers").orderByChild(str).endAt(str + "~"), MainModel.class)
                        .build();
        mainAdapter = new MainAdapter(options);
        mainAdapter.startListening();
        recyclerView.setAdapter(mainAdapter);
    }
}

File: 4/android/temu3/app/src/main/java/com/example/temu3/MainActivity.kt
package com.example.temu3

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.example.temu3.references.GreetingActivity
import com.example.temu3.references.GuessNumberActivity
import com.example.temu3.references.HelloWorldActivity
import com.example.temu3.references.PartsListActivity
import com.example.temu3.references.SmartHomeClientActivity
import com.example.temu3.temu3.HitungUmurActivity
import com.example.temu3.temu3.KebunBinatangActivity
import com.example.temu3.temu3.SquadSoccerActivity
import com.example.temu3.temu4.HitungEmasActivity
import com.example.temu3.temu5.ToastDemoActivity
import com.example.temu3.temu5.WebViewActivity


class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        // Pertemuan 3 buttons
        val kebunBinatangButton = findViewById<Button>(R.id.kebun_binatang)
        kebunBinatangButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                KebunBinatangActivity::class.java
            )
            startActivity(intent)
        }
        val squadSoccerButton = findViewById<Button>(R.id.squad_soccer)
        squadSoccerButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                SquadSoccerActivity::class.java
            )
            startActivity(intent)
        }
        val hitungUmurButton = findViewById<Button>(R.id.hitung_umur)
        hitungUmurButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                HitungUmurActivity::class.java
            )
            startActivity(intent)
        }

        // Pertemuan 4 buttons
        val hitungEmasButton = findViewById<Button>(R.id.hitung_emas)
        hitungEmasButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                HitungEmasActivity::class.java
            )
            startActivity(intent)
        }

        // Pertemuan 5 buttons
        val webViewTPLButton = findViewById<Button>(R.id.webview_tpl)
        webViewTPLButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                WebViewActivity::class.java
            )
            startActivity(intent)
        }
        val toastDemoButton = findViewById<Button>(R.id.toast_demo)
        toastDemoButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                ToastDemoActivity::class.java
            )
            startActivity(intent)
        }

        // References buttons
        val helloWorldButton = findViewById<Button>(R.id.hello_world)
        helloWorldButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                HelloWorldActivity::class.java
            )
            startActivity(intent)
        }
        val guessNumberButton = findViewById<Button>(R.id.guess_number)
        guessNumberButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                GuessNumberActivity::class.java
            )
            startActivity(intent)
        }
        val greetingAppButton = findViewById<Button>(R.id.greeting_app)
        greetingAppButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                GreetingActivity::class.java
            )
            startActivity(intent)
        }
        val smartHomeClientButton = findViewById<Button>(R.id.smart_home_client)
        smartHomeClientButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                SmartHomeClientActivity::class.java
            )
            startActivity(intent)
        }
        val partsListButton = findViewById<Button>(R.id.parts_list)
        partsListButton.setOnClickListener {
            val intent = Intent(
                this@MainActivity,
                PartsListActivity::class.java
            )
            startActivity(intent)
        }
    }
}
